Gems/Atom/RHI/Code/Source/RHI.Edit/ShaderBuildArguments.cpp

```diff
@@ -0,0 +1,126 @@
+                    ->Field("cpp", &ShaderBuildArguments::m_cppArguments)
+                    ->Field("azslc", &ShaderBuildArguments::m_azslcArguments)
+                    ->Field("dxc", &ShaderBuildArguments::m_dxcArguments)
+                    ->Field("spirv-cross", &ShaderBuildArguments::m_spirvCrossArguments)
+                    ->Field("metalair", &ShaderBuildArguments::m_metalAirArguments)
+                    ->Field("metallib", &ShaderBuildArguments::m_metalLibArguments)
+                    ;
+            }
+        }
+
+        ShaderBuildArguments::ShaderBuildArguments(
+                bool generateDebugInfo,
+                const AZStd::vector<AZStd::string>& cppArguments,
+                const AZStd::vector<AZStd::string>& azslcArguments,
+                const AZStd::vector<AZStd::string>& dxcArguments,
+                const AZStd::vector<AZStd::string>& spirvCrossArguments,
+                const AZStd::vector<AZStd::string>& metalAirArguments,
+                const AZStd::vector<AZStd::string>& metalLibArguments)
+            : m_generateDebugInfo(generateDebugInfo)
+            , m_cppArguments(cppArguments)
+            , m_azslcArguments(azslcArguments)
+            , m_dxcArguments(dxcArguments)
+            , m_spirvCrossArguments(spirvCrossArguments)
+            , m_metalAirArguments(metalAirArguments)
+            , m_metalLibArguments(metalLibArguments)
+        {
+        }
+
+        bool ShaderBuildArguments::HasArgument(const AZStd::vector<AZStd::string>& argList, const AZStd::string& arg)
+        {
+            return AZStd::find(argList.begin(), argList.end(), arg) != argList.end();
+        }
+
+        void ShaderBuildArguments::AppendArguments(AZStd::vector<AZStd::string>& out, const AZStd::vector<AZStd::string>& in)
+        {
+            for (const auto& argument : in)
+            {
+                if (!HasArgument(out, argument))
+                {
+                    out.push_back(argument);
+                }
+            }
+        }
+
+        void ShaderBuildArguments::RemoveArguments(AZStd::vector<AZStd::string>& out, const AZStd::vector<AZStd::string>& in)
+        {
+            out.erase(remove_if(out.begin(), out.end(),
+                [&](const AZStd::string& arg)
+                {
+                    return HasArgument(in, arg);
+                }), std::end(out));
+        }
+
+        AZStd::string ShaderBuildArguments::ListAsString(const AZStd::vector<AZStd::string>& argList)
+        {
+            AZStd::string listAsString;
+            AZ::StringFunc::Join(listAsString, argList, " ");
+            return listAsString;
+        }
+
+        ShaderBuildArguments ShaderBuildArguments::operator+(const ShaderBuildArguments& rhs) const
+        {
+            auto buildArguments = *this;
+            buildArguments += rhs;
+            return buildArguments;
+        }
+
+        ShaderBuildArguments& ShaderBuildArguments::operator+=(const ShaderBuildArguments& rhs)
```

Comment:
It's a little strange that the shader build argument type doesn't understand which node in the compiler pipeline it corresponds to.

Reply:
Well, it has dedicated fields for each tool. But not sure what you mean.